### DISCUSSION TITLE
What does the executable do?

### DISCUSSION THREAD

**Discussion by aDramaQueen** [django/daphne#524] [Q&A] [answered]
Hi,
my goal is to pass a configuration file to Daphne and then use it to start the server. The configuration file should be written in JSON. Since parsing a JSON file via shell scripts is extremely tedious, I wanted to do this in Python as well. I then discovered that I can also start the Daphne CLI from Python.
Now the question: The executable from PiPy (depending on your system) only starts the CLI? Or is there more going on? In other words, could the following Python code be considered equivalent:

```python
def _parse_json() -> list[str | None]:
    # TODO: Parse JSON
    return []


def _delete_pid_file(path: Path):
    path.unlink(missing_ok=True)


def _pid_file() -> None:
    pid_file = Path(os.getcwd(), 'daphne.pid')

    # Write PID file
    with open(pid_file, 'w', encoding='utf-8') as file:
        file.write(str(os.getpid()))

    # Make sure PID file is deleted, if the Python interpreter stops working
    # @see https://docs.python.org/3.12/library/atexit.html
    register(_delete_pid_file, pid_file)


if __name__ == '__main__':
    print('Starting ASGI server...')

    _pid_file()

    cli = CommandLineInterface()
    cli.run(_parse_json())

    print('Stopping ASGI server, bye bye...')
```

**Comment by carltongibson** [comment, marked as answer]
> The executable from PiPy (depending on your system) only starts the CLI? 

Yep. Exactly.

https://github.com/django/daphne/blob/4a55faca45db2e2a4c275ee02c44893e6af0b5c7/daphne/cli.py#L204
